ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs

```diff
@@ -177,6 +177,13 @@ instance
     LedgerSupportsProtocol (ShelleyBlock proto era)
   ) => SingleEraBlock (ShelleyBlock proto era) where
   singleEraTransition pcfg _eraParams _eraStart ledgerState =
+      -- TODO(jdral): When fixing a bug in @db-analyser@ related to hard-forks,
+      -- we replaced the matched expression below with @traceShowId $
+      -- shelleyTriggerHardFork pcf@ to debug the hard-fork functionality.
+      -- The trace of @db-analyser@ shows that the expression is printed thrice
+      -- for every @reapplyBlock@ of a Shelley block. Does this mean that we
+      -- evaluate this function thrice for every @reapplyBlock@? It would be
+      -- useful to investigate whether this is problematic behaviour.
```

Comment:
This will print something along the lines of the following:
```
TriggerHardForkAtVersion 3
TriggerHardForkAtVersion 3
TriggerHardForkAtVersion 3
```

Reply:
I've created ticket CAD-4666, which also links to the updated TODO in the code.